janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeyIterator.java

```diff
@@ -34,6 +35,9 @@
      * depending on the implementation, hence it is important to iterate over
      * (and close) the RecordIterator before calling {@link #next()} or {@link #hasNext()}.
      *
+     * Important! Entries should be sorted inside iterator.
```

Comment:
Isn't the iterator itself ought to be sorted? From my understanding, the entries returned by the "getEntries()" don't have to be sorted.

Reply:
It's critical for `VertexJobConverter.isGhostVertex`. This method checks only first entry of each row
```
protected boolean isGhostVertex(long vertexId, EntryList firstEntries) {
        if (idManager.isPartitionedVertex(vertexId) && !idManager.isCanonicalVertexId(vertexId)) return false;

        RelationCache relCache = tx.getEdgeSerializer().parseRelation(
                firstEntries.get(0),true,tx);
        return relCache.typeId != BaseKey.VertexExists.longId();
    }
```

Reply:
Sorry I was wrong by saying "entries returned by the "getEntries()" don't have to be sorted". In fact, it has to be sorted because of the [JanusGraph Data Model](https://docs.janusgraph.org/advanced-topics/data-model/): "The cells must be sorted by their columns and a subset of the cells specified by a column range must be efficiently retrievable". Therefore this notice seems a bit redundant to me.

Reply:
For me it was not obvious. Spent a decent amount of time trying to figure out why test is not working. And the root cause was wrong order of entries in row. If you insist I may remove this note. At least I've fixed my test.

Reply:
I'd rather retain comment and add link that you provided

Reply:
I think I missed something earlier. When I saw this PR, I presumed your data backend does sort columns natively but does not sort keys. That made perfect sense to me because apparently, the StandardScannerExcecutor assumed key orders are "consistent" (either sorted by value like ByteOrderedPartitioner in Cassandra or sorted by key's hash value like Murmur3Partitioner in Cassandra) among multiple scans.

Now, seems your data backend neither supports key order nor supports column order. In other words, rows are not sorted in any way (neither globally nor at partition level). Then it makes perfect sense why you add this comment because your data backend actually does not support that natively. Am I understanding it correctly? If yes, then it would be very interesting to me, because AFAIK some features actually rely on that assumption (columns are sorted) - that is a bit off-topic, though.

Reply:
Exactly. Need to manually sort column values before iterating on them

Reply:
Actually Aerospike is a KeyValue store but it allows to emulate KeyColumnValue store by using Map structure in place of value.

BTW, after this fix (KeySlicesIterator introduced) all Aerospike tests become green.

Reply:
Got it, thanks. Then it would be good to keep this comment. I'll just do another pass before approving this PR.

At the same time, I am just curious: If columns are unsorted, vertex-centric indexes will be useless when it comes to range queries (e.g. `g.V(h).outE('battled').has('time', inside(10, 20)).inV()` as shown in https://docs.janusgraph.org/index-management/index-performance/#vertex-centric-indexes) because this feature relies on the columns to be sorted. Have you used vertex-centric indexes and if so, what performance have you observed? I am asking because if not supporting column order actually does not affect normal use (despite might be inefficient in some scenarios), we should update the data model documentation - currently, it says "The cells must be sorted by their columns".

Reply:
Aerospike connector supports all types of indexes (vertex, edge) and we used this feature on several production services.
The main reason is that Aerospike implementation of this method `KeyColumnValueStore.getSlice(KeySliceQuery query, StoreTransaction txh)` returns sorted entries.
WDYT about put this comment regarding column order on EntryList interface also?

Reply:
> The main reason is that Aerospike implementation of this method KeyColumnValueStore.getSlice(KeySliceQuery query, StoreTransaction txh) returns sorted entries.

Got it. I can see the usage of MapOperation::getByKeyRange which supports range query. Earlier I thought you had to retrieve the whole Map to do in-memory filtering for "KeyColumnValueStore.getSlice(KeySliceQuery query, StoreTransaction txh)" calls.

> WDYT about put this comment regarding column order on EntryList interface also

At this moment, I am not 100% sure whether "EntryList getSlice(KeySliceQuery query, StoreTransaction txh)" requires the ordering of columns. Column ordering has been a fundamental requirement but now it looks to me that this is not a hard requirement. I would suggest not doing this right now until we have experiment results showing that the EntryList interface must be sorted by column.

Reply:
The final proposal
WDYT: RecordsCollector  -> RowsCollector ?
It actually collects rows to queue.

Reply:
done